dts/bindings/display/sitronix,st7796s.yaml

```diff
@@ -104,3 +104,12 @@ properties:
       - "2-dot"
     description: |
       Display inversion control mode.
+
+  rgb-is-inverted:
+    type: boolean
+    description: |
+      Inverting color format order (RGB->BGR or BGR->RGB)
+      In the case of enabling this option, API reports pixel-format in capabilities
+      as the inverted value of the RGB pixel-format specified in MADCTL.
+      This option is convenient for supporting displays with bugs
```

Comment:
displays with bugs? what does this mean?

Reply:
DIsplays that have the behave as BGR instead of when requesting RGB and vice versa. This is a similar to the https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/display/sitronix%2Cst7735r.yaml#L127C5-L133C75

Reply:
Hello @mmahadevan108 , @decsny ,

Although the description here and the comment in st7735 driver says it is display controller's bug, I have different understanding.

https://github.com/zephyrproject-rtos/zephyr/blob/4efc03520966a96e1a5dcf5369f75503626c2b42/drivers/display/display_st7735r.c#L276

Firstly, for RGB565&BGR565, the LCD controller's pixel format name is different with Zephyr:
1. The LCD controller setting ST7796S_MADCTL_BGR is to swap color channels R and B. As shown in the LCD controller's document:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/31212062/4aee7529-bb26-420a-8965-255f3a876cb4)
2. Zephyr's RGB565 & BGR565 is byte swap but not color channel swap.
PIXEL_FORMAT_RGB_565
![image](https://github.com/zephyrproject-rtos/zephyr/assets/31212062/3e260f69-6031-4b38-878e-632512717a4b)
PIXEL_FORMAT_BGR_565
![image](https://github.com/zephyrproject-rtos/zephyr/assets/31212062/9822367f-d16f-4c11-b35b-87c30e0b2d1d)
We can get this from the function get_rgb565_color, fill_buffer_rgb565, fill_buffer_bgr565 in display driver's sample:
https://github.com/zephyrproject-rtos/zephyr/blob/4efc03520966a96e1a5dcf5369f75503626c2b42/samples/drivers/display/src/main.c#L98-L141
In short, Zephyr's RGB565 and BGR565 are just ST7796S's RGB565, in different endian modes. So ST7796S_MADCTL_BGR shall always be set.

Secondly, the transfer interface byte order needs be considered. There are SPI, I8080-8bit, I8080-16bit.
These are the ST7796S required transfer formats:
I8080-16bit
![image](https://github.com/zephyrproject-rtos/zephyr/assets/31212062/17adf2a4-1d93-42a4-a24a-bb22b762ee02)
I8080-8bit
![image](https://github.com/zephyrproject-rtos/zephyr/assets/31212062/453ec02a-6bf2-4f65-aa48-514e19659b40)
SPI
![image](https://github.com/zephyrproject-rtos/zephyr/assets/31212062/a73d1535-1707-4ba6-b20d-1703dd5c167d)

If the pixels in memory is PIXEL_FORMAT_BGR_565
![image](https://github.com/zephyrproject-rtos/zephyr/assets/31212062/9822367f-d16f-4c11-b35b-87c30e0b2d1d)

With I8080-16bit, flexio LCD driver sends low byte to D0-D7, high byte to D8-D16, it matches the ST7796S requirement. PIXEL_FORMAT_BGR_565 can be used.
With I8080-8bit and SPI, low byte is sent first, it doesn't match the ST7796S requirement, the pixel byte need be swapped. PIXEL_FORMAT_RGB_565 shall be used.

LVGL's RGB565 is actually PIXEL_FORMAT_BGR_565, if PIXEL_FORMAT_RGB_565 is desired, LV_COLOR_16_SWAP need be set.

In summary:
| Interface | Pixel Format | LV_COLOR_16_SWAP |
| ------------- | ------------- | ------------- |
| I8080-16bit  | PIXEL_FORMAT_BGR_565  | n |
| I8080-8bit/SPI  | PIXEL_FORMAT_RGB_565  | y |

BTW, IMHO PIXEL_FORMAT_BGR_565 is not a good naming, because generally RGB/BGR is color channel swap.

Reply:
Hi @zejiang0jason, thank you for this excellent explanation. First off, I agree that the `PIXEL_FORMAT_BGR_565` is poor naming on Zephyr's part. I see it is as a historical choice we simply have to live with (as changing the API would likely cause breakage downstream). We could (and likely should) improve the documentation around these values in the display API though.


Based on this, it seems like we should be reporting `RGB` and `BGR` format based on the *interface* in use, not the setting of the MADCTL bit. Would you say this is correct? If so, we probably need different MIPI mode definitions for 8 bit versus 16 bit 8080 interfaces

Reply:
Hi @danieldegrasse , yes, I agree that the format needs base on the _interface_ used. And we need to clearly define the byte order on the interface, because some IPs are very powerful, they support different endian modes (byte swap). For example, the LCDIC (on RW612), it is 8-bit 8080, but supports sending both low-byte first and high byte first.

Reply:
> For example, the LCDIC (on RW612), it is 8-bit 8080, but supports sending both low-byte first and high byte first.

Sure- we support this in Zephyr currently (at least for SPI mode). It is primarily used for LVGL, where we can skip the overhead of setting `LV_COLOR_16_SWAP`. I'm not sure how we'd allow the display driver to determine that the interface was swapping bytes. We could use the `rgb-is-inverted` property in this case, so that the display reports a byte swapped color format (this way the color format could be swapped by the LCDIC or other IP)

Reply:
@danieldegrasse @zejiang0jason , thank you for your help with suggestions and testing. I have update the `st7796s.c` to add checks for 8080 8-bit mode. I have also tested with 8-bit and 16-bit data bus width.

Reply:
Do we still need support for the `rgb-is-inverted` property given the discussion here? Or can it be dropped?

If we have a use case I think adding support for this makes sense, but otherwise I would lean towards removing it to minimize the size of the driver

Reply:
This is not needed for the display we are using on the MCX boards. However might be required for other displays.